src/gui/plugins/environment_visualization/EnvironmentVisualization.cc

```diff
@@ -0,0 +1,312 @@
+      auto dy = step.Y() / ySamples;
+      auto dz = step.Z() / zSamples;
+      std::size_t idx = 0;
+      for (std::size_t x_steps = 0; x_steps < ceil(xSamples); x_steps++)
+      {
+        auto x = lower_bound.X() + x_steps * dx;
+        for (std::size_t y_steps = 0; y_steps < ceil(ySamples); y_steps++)
+        {
+          auto y = lower_bound.Y() + y_steps * dy;
+          for (std::size_t z_steps = 0; z_steps < ceil(zSamples); z_steps++)
+          {
+            auto z = lower_bound.Z() + z_steps * dz;
+            auto res = frame.LookUp(
+              session, math::Vector3d(x, y, z));
+
+            if (res.has_value())
+            {
+              this->floatFields[key].mutable_data()->Set(idx,
+                static_cast<float>(res.value()));
+            }
+            idx++;
+          }
+        }
+      }
+    }
+  }
+
+  /////////////////////////////////////////////////
+  public: void Publish()
+  {
+    pcPub.Publish(this->pcMsg);
+    for(auto &[key, pub] : this->pubs)
+    {
+      pub.Publish(this->floatFields[key]);
+    }
+  }
+
+  /////////////////////////////////////////////////
+  public: void ResizeCloud(
+    std::shared_ptr<components::EnvironmentalData> data,
+    const EntityComponentManager& _ecm,
+    double xSamples, double ySamples, double zSamples)
+  {
+    assert (pubs.size() > 0);
+
+    // Assume all data have same point cloud.
+    gz::msgs::InitPointCloudPacked(pcMsg, "some_frame", true,
+        {{"xyz", gz::msgs::PointCloudPacked::Field::FLOAT32}});
+    auto numberOfPoints =
+      ceil(xSamples) * ceil(ySamples) * ceil(zSamples);
+    std::size_t dataSize{numberOfPoints * pcMsg.point_step()};
+    pcMsg.mutable_data()->resize(dataSize);
+    pcMsg.set_height(1);
+    pcMsg.set_width(numberOfPoints);
+
+    auto session = this->sessions[this->pubs.begin()->first];
+    auto frame = data->frame[this->pubs.begin()->first];
+    auto [lower_bound, upper_bound] =
+      frame.Bounds(session);
+
+    auto step = upper_bound - lower_bound;
+    auto dx = step.X() / xSamples;
+    auto dy = step.Y() / ySamples;
+    auto dz = step.Z() / zSamples;
+
+    // Populate point cloud
+    gz::msgs::PointCloudPackedIterator<float> xIter(pcMsg, "x");
+    gz::msgs::PointCloudPackedIterator<float> yIter(pcMsg, "y");
+    gz::msgs::PointCloudPackedIterator<float> zIter(pcMsg, "z");
+
+    for (std::size_t x_steps = 0; x_steps < ceil(xSamples); x_steps++)
+    {
+      for (std::size_t y_steps = 0; y_steps < ceil(ySamples); y_steps++)
+      {
+        for (std::size_t z_steps = 0; z_steps < ceil(zSamples); z_steps++)
+        {
+          auto x = lower_bound.X() + x_steps * dx;
+          auto y = lower_bound.Y() + y_steps * dy;
+          auto z = lower_bound.Z() + z_steps * dz;
+
+          auto coords = GetGridFieldCoordinates(
+            _ecm, math::Vector3d{x, y, z},
+            data);
+
+          if (!coords.has_value())
+          {
+            return;
+          }
```

Comment:
@arjo129 doesn't this mean there won't be a 1-to-1 match between float fields and cloud points?